File: src/frontend/src/api/assessments.ts
import { apiClient } from './client'
import type {
  CampaignRead,
  CampaignCreate,
  AssessmentRead,
  AssessmentCreate,
  AssessmentScoreSubmit,
  CampaignProgressRead,
  AggregatedScoreRead,
} from '../types/assessment'

export const assessmentsApi = {
  listCampaigns: (status?: string) =>
    apiClient
      .get<CampaignRead[]>('/assessments/campaigns', { params: { status } })
      .then((r) => r.data),

  getCampaign: (id: string) =>
    apiClient.get<CampaignRead>(`/assessments/campaigns/${id}`).then((r) => r.data),

  createCampaign: (data: CampaignCreate) =>
    apiClient.post<CampaignRead>('/assessments/campaigns', data).then((r) => r.data),

  createAssessment: (data: AssessmentCreate) =>
    apiClient.post<AssessmentRead>('/assessments', data).then((r) => r.data),

  listAssessments: (params?: { campaign_id?: string; assessee_id?: string }) =>
    apiClient.get<AssessmentRead[]>('/assessments', { params }).then((r) => r.data),

  getAssessment: (id: string) =>
    apiClient.get<AssessmentRead>(`/assessments/${id}`).then((r) => r.data),

  submitScores: (id: string, data: AssessmentScoreSubmit) =>
    apiClient.post<AssessmentRead>(`/assessments/${id}/scores`, data).then((r) => r.data),

  activateCampaign: (id: string) =>
    apiClient.post<CampaignRead>(`/assessments/campaigns/${id}/activate`).then((r) => r.data),

  closeCampaign: (id: string) =>
    apiClient.post<CampaignRead>(`/assessments/campaigns/${id}/close`).then((r) => r.data),

  finalizeCampaign: (id: string) =>
    apiClient.post<CampaignRead>(`/assessments/campaigns/${id}/finalize`).then((r) => r.data),

  archiveCampaign: (id: string) =>
    apiClient.post<CampaignRead>(`/assessments/campaigns/${id}/archive`).then((r) => r.data),

  getCampaignProgress: (id: string) =>
    apiClient.get<CampaignProgressRead>(`/assessments/campaigns/${id}/progress`).then((r) => r.data),

  getAggregatedScores: (id: string) =>
    apiClient.get<AggregatedScoreRead[]>(`/assessments/campaigns/${id}/scores`).then((r) => r.data),

  setWeights: (id: string, weights: { department_head_weight: number; team_lead_weight: number; self_weight: number; peer_weight: number }) =>
    apiClient.put(`/assessments/campaigns/${id}/weights`, weights).then((r) => r.data),

  setPeers: (campaignId: string, peerIds: string[]) =>
    apiClient.post(`/assessments/campaigns/${campaignId}/peers`, { peer_ids: peerIds }).then((r) => r.data),

  getMyPeers: (campaignId: string) =>
    apiClient.get<string[]>(`/assessments/campaigns/${campaignId}/peers`).then((r) => r.data),

  listMyTasks: () =>
    apiClient.get<AssessmentRead[]>('/assessments/my-tasks').then((r) => r.data),
}

File: src/backend/app/api/assessments.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import get_current_user, require_roles
from app.models.enums import CampaignStatus
from app.models.user import User
from app.schemas.assessment import (
    AssessmentCreate,
    AssessmentRead,
    AssessmentScoreSubmit,
    CampaignCreate,
    CampaignRead,
)
from app.services.assessment import AssessmentService

router = APIRouter(prefix="/assessments", tags=["assessments"])

ERROR_MAP = {
    "not_found": (status.HTTP_404_NOT_FOUND, "Не найдено"),
    "duplicate": (status.HTTP_409_CONFLICT, "Оценка уже существует"),
    "invalid_dates": (status.HTTP_400_BAD_REQUEST, "Дата окончания должна быть позже даты начала"),
}


def _raise(code: str) -> None:
    status_code, detail = ERROR_MAP.get(code, (400, code))
    raise HTTPException(status_code=status_code, detail=detail)


@router.get("/campaigns", response_model=list[CampaignRead])
async def list_campaigns(
    campaign_status: CampaignStatus | None = Query(default=None, alias="status"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = AssessmentService(db)
    return await service.list_campaigns(current_user, campaign_status)


@router.post(
    "/campaigns",
    response_model=CampaignRead,
    status_code=status.HTTP_201_CREATED,
)
async def create_campaign(
    data: CampaignCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_roles("admin", "head", "department_head")),
):
    service = AssessmentService(db)
    try:
        campaign = await service.create_campaign(data, current_user)
    except ValueError as e:
        _raise(str(e))
    return campaign


@router.get("/campaigns/{campaign_id}", response_model=CampaignRead)
async def get_campaign(
    campaign_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    service = AssessmentService(db)
    try:
        campaign = await service.get_campaign(campaign_id)
    except ValueError as e:
        _raise(str(e))
    return campaign


@router.post("", response_model=AssessmentRead, status_code=status.HTTP_201_CREATED)
async def create_assessment(
    data: AssessmentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_roles("admin", "team_lead", "department_head")),
):
    service = AssessmentService(db)
    try:
        assessment = await service.create_assessment(data, current_user)
    except ValueError as e:
        _raise(str(e))
    return assessment


@router.get("", response_model=list[AssessmentRead])
async def list_assessments(
    campaign_id: uuid.UUID | None = Query(default=None),
    assessee_id: uuid.UUID | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = AssessmentService(db)
    return await service.list_assessments(current_user, campaign_id, assessee_id)


@router.get("/{assessment_id}", response_model=AssessmentRead)
async def get_assessment(
    assessment_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    service = AssessmentService(db)
    try:
        assessment = await service.get_assessment(assessment_id)
    except ValueError as e:
        _raise(str(e))
    return assessment


@router.post("/{assessment_id}/scores", response_model=AssessmentRead)
async def submit_scores(
    assessment_id: uuid.UUID,
    data: AssessmentScoreSubmit,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    service = AssessmentService(db)
    try:
        assessment = await service.submit_scores(assessment_id, data)
    except ValueError as e:
        _raise(str(e))
    return assessment
